Code/Source/Camera/ROS2CameraSensorComponent.cpp

```diff
@@ -121,39 +123,59 @@ namespace ROS2
         {
             return;
         }
+
+        if (!frames_publish_queue->empty())
+        {
+            ProcessQueue();
+        }
+
         m_cameraSensor->RequestFrame(
             transform,
-            [this](const AZ::RPI::AttachmentReadback::ReadbackResult& result)
+            [task_queue = frames_publish_queue](const AZ::RPI::AttachmentReadback::ReadbackResult& result)
             {
-                const AZ::RHI::ImageDescriptor& descriptor = result.m_imageDescriptor;
-                const auto* component = Utils::GetGameOrEditorComponent<ROS2FrameComponent>(GetEntity());
-                AZStd::string frameName = component->GetFrameID();
-                sensor_msgs::msg::Image message;
-                message.encoding = sensor_msgs::image_encodings::RGBA8;
-
-                message.width = descriptor.m_size.m_width;
-                message.height = descriptor.m_size.m_height;
-                message.step = message.width * sensor_msgs::image_encodings::bitDepth(message.encoding) / 8 *
-                    sensor_msgs::image_encodings::numChannels(message.encoding);
-                message.data = std::vector<uint8_t>(result.m_dataBuffer->data(), result.m_dataBuffer->data() + result.m_dataBuffer->size());
-                message.header.frame_id = frameName.c_str();
-
-                m_imagePublisher->publish(message);
-
-                sensor_msgs::msg::CameraInfo cameraInfo;
-                cameraInfo.header.frame_id = frameName.c_str();
-                cameraInfo.width = descriptor.m_size.m_width;
-                cameraInfo.height = descriptor.m_size.m_height;
-                cameraInfo.distortion_model = sensor_msgs::distortion_models::PLUMB_BOB;
-                AZ_Assert(m_cameraSensor->GetCameraSensorDescription().m_cameraIntrinsics.size() == cameraInfo.k.size(), "should be 9");
-                std::copy_n(
-                    m_cameraSensor->GetCameraSensorDescription().m_cameraIntrinsics.data(),
-                    m_cameraSensor->GetCameraSensorDescription().m_cameraIntrinsics.size(),
-                    cameraInfo.k.begin());
-                cameraInfo.p = { cameraInfo.k[0], cameraInfo.k[1], cameraInfo.k[2], 0, cameraInfo.k[3], cameraInfo.k[4], cameraInfo.k[5], 0,
-                                 cameraInfo.k[6], cameraInfo.k[7], cameraInfo.k[8], 0 };
-
-                m_cameraInfoPublisher->publish(cameraInfo);
+                task_queue->push(FrameTask{ ROS2Interface::Get()->GetROSTimestamp(),
```

Comment:
Why there is no mutex or other synchronization here? We are modifying the container.